### PR TITLE
Issue #SB-8462 fix: Streaming video while loading in content player

### PR DESCRIPTION
**Pull request by gouravmore**
**Issue :-** https://project-sunbird.atlassian.net/browse/SB-5992
**Scenarios:-** https://project-sunbird.atlassian.net/browse/SB-8470

-----------------------------------------------------------------------------------------------------
**No. of Functional test cases:-** _16_
**No. of Unit test case:-** _6_

-----------------------------------------------------------------------------------------------------
**Tested scenarios:-** 

*_Video Streaming_*
- [x] While playing the live streaming of content in a content player, the entire video should not load.
- [x] While playing the live streaming of content in a content player, It has to stream from the server.
- [x] On slow/low network - the live streaming video has to be shown

_*Player behavior*_
- [x] MP4 video has to play in the content player
- [x] WebM video has to play in the content player
- [x] the youtube video has to play in the content player
- [x] The video should be able to play and pause at any movement of a video
- [x] The video should be able to mute and unmute at any movement of the video

_*Download video*_
- [x] MP4 and WebM video should be able to download
- [x] Stream and youtube video should not be able to download

_*Telemetry should have been logged for following events*_
- [x] START - on start of video
- [x] END - on end of video
- [x] INTERACT - on Seeked video
- [x] INTERACT - on play/pause

_*Error handling*_
- [x] If streaming URL is wrong, the error message has to be shown(!)
- [x] If there is no internet connectivity, the error message has to be shown (!) [Message = Internet not available. Please connect and try again.]

### CODE REVIEW

File: player/public/coreplugins/org.ekstep.videorenderer-1.0/renderer/plugin.js [hunk 1]
```diff
@@ -37,30 +39,76 @@ org.ekstep.contentrenderer.baseLauncher.extend({
...
+    _loadVideo: function(path, data) {
+        var instance = this;
+        if (instance.isStreaming && !navigator.onLine) {
+            EkstepRendererAPI.logErrorEvent('No internet', {
+                'type': 'content',
+                'action': 'play',
+                'severity': 'error'
+            });
+            instance.throwError({message:'Internet not available. Please connect and try again.'});
+        }
```
Comment (vinukumar-vs) [inline player/public/coreplugins/org.ekstep.videorenderer-1.0/renderer/plugin.js]: break the function execution, as noInternet not need to execute any further lines for streaming.

File: player/public/coreplugins/org.ekstep.videorenderer-1.0/renderer/plugin.js [hunk 2]
```diff
@@ -37,30 +39,76 @@ org.ekstep.contentrenderer.baseLauncher.extend({
...
         var source = document.createElement("source");
         source.src = path;
+        source.type = data.mimeType;
         video.appendChild(source);
-        this.addvideoListeners(video);
-        this.videoPlayer = video;
+
+        if (instance.isStreaming || window.cordova){
+            var videoPlayer = videojs('videoElement', {
+                "controls": true, "autoplay": true, "preload": "auto"
```
Comment (vinukumar-vs) [inline player/public/coreplugins/org.ekstep.videorenderer-1.0/renderer/plugin.js]: Why autoPlay & controls by default is true?

It should be based on property of video element autoplay="true/false".

File: player/public/coreplugins/org.ekstep.videorenderer-1.0/renderer/plugin.js [hunk 3]
```diff
@@ -146,25 +194,35 @@ org.ekstep.contentrenderer.baseLauncher.extend({
...
     },
-    addvideoListeners: function(videoPlayer) {
+    addVideoListeners: function(videoPlayer, path) {
         var instance = this;
 
-        videoPlayer.onplay = function(e) {
-            instance.play("videostage", Math.floor(instance.videoPlayer.timeStamp)*1000);
-        };
+        videoPlayer.on("play", function(e) {
+            instance.play("videostage", Math.floor(instance.videoPlayer.currentTime())*1000);
```
Comment (vinukumar-vs) [inline player/public/coreplugins/org.ekstep.videorenderer-1.0/renderer/plugin.js]: Why did you changed timeStamp to currentTime()?

File: player/public/coreplugins/org.ekstep.videorenderer-1.0/renderer/plugin.js [hunk 4]
```diff
@@ -146,25 +194,35 @@ org.ekstep.contentrenderer.baseLauncher.extend({
...
-        videoPlayer.onseeked = function(e) {
-            instance.seeked("videostage", Math.floor(instance.videoPlayer.timeStamp)*1000);
-        };
+        videoPlayer.on("seeked", function(e) {
+            instance.seeked("videostage", Math.floor(instance.videoPlayer.currentTime())*1000);
+        });
 
+        if (instance.isStreaming){
+            videoPlayer.on("error", function(e) {
+                EventBus.dispatch("renderer:alert:show", undefined, {
```
Comment (vinukumar-vs) [inline player/public/coreplugins/org.ekstep.videorenderer-1.0/renderer/plugin.js]: There is no message to the user?

File: player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+            if (typeof cordova !== "undefined") exitApp();
+            return false;
+        }
+        var source = document.createElement("source");
+        source.src = path;
+        source.type = data.mimeType;
+        video.appendChild(source);
+
+        if (instance.isStreaming || window.cordova){
+            var videoPlayer = videojs('videoElement', {
```
Comment (manjudr) [inline player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js]: Use try catch block and handle the error

File: player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+    _loadVideo: function(path, data) {
+        var instance = this;
+        if (instance.isStreaming && !navigator.onLine) {
+            EkstepRendererAPI.logErrorEvent('No internet', {
+                'type': 'content',
+                'action': 'play',
+                'severity': 'error'
+            });
+            instance.throwError({message: instance.messages.noInternetConnection});
+            if (typeof cordova !== "undefined") exitApp();
```
Comment (manjudr) [inline player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js]: why are doing force `exitApp();` ?

Comment (manjudr) [reply player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js]: exitApp is not required i guess. Either it should re-connect to internet and start play video or let user to exit the player.

File: player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+                vjsdownload:{
+                beforeElement: 'playbackRateMenuButton',
+                textControl: 'Download video',
+                name: 'downloadButton'
+                }
+            }
+            } , function() {
+                this.on('downloadvideo', function(){
+                    EkstepRendererAPI.getTelemetryService().interact("TOUCH", "Download", "TOUCH", {
+                        stageId: 'videostage',
```
Comment (manjudr) [inline player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js]: stageId is constant

File: player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    },
+    currentTime: 1,
+    videoPlayer: undefined,
+    stageId: undefined,
+    heartBeatData:{},
+    enableHeartBeatEvent:false,
+    _constants: {
+        mimeType: ["video/mp4", "video/x-youtube", "video/webm"],	
+        events: {	
+            launchEvent: "renderer:launch:video"
```
Comment (manjudr) [inline player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js]: why single event is defined here? what about other events ?`"renderer:overlay:mute"`

File: player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+        this.configOverlay();
+    },
+    createVideo: function(path, data) {
+        EkstepRendererAPI.dispatchEvent("renderer:splash:hide");
+
+        video = document.createElement('video-js');
+        video.style.width = '100%';
+        video.style.height = '100%';
+        video.style.position = 'absolute';
+        video.id = "videoElement";
```
Comment (manjudr) [inline player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js]: don't use generic-name as a identifier `videoElement ` better you should user `this.manifest.id` as a video element identifier

File: player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+            this._loadVideo(data.previewUrl, data);
+        }else{
+            this._loadVideo(path, data);
+        }
+        
+        video.onvolumechange = function() {
+            (video.muted) ? EkstepRendererAPI.dispatchEvent('renderer:overlay:mute') : EkstepRendererAPI.dispatchEvent('renderer:overlay:unmute');
+        };
+        $("video").bind("contextmenu",function() {
+            return false;
```
Comment (manjudr) [inline player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js]: Im unclear with `contextmenu` what it is doing? if it is not required please remove this

File: player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+            var youtubeInstance = this;
+            youtubeInstance.src({
+                type: 'video/youtube',
+                src: path
+            });
+            instance.addYOUTUBEListeners(youtubeInstance);
+            instance.setYoutubeStyles(youtubeInstance);
+            instance.videoPlayer = youtubeInstance;
+            /*youtubeInstance.on('volumechange', function(){
+                (youtubeInstance.muted()) ? EkstepRendererAPI.dispatchEvent('renderer:overlay:mute') : EkstepRendererAPI.dispatchEvent('renderer:overlay:unmute');
```
Comment (manjudr) [inline player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js]: remove the comments code

File: player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+            });*/            
+        });
+    },
+    setYoutubeStyles: function(youtube) {
+        var instance = this;
+        videojs("videoElement").ready(function() {
+            var video = document.getElementById("videoElement");
```
Comment (manjudr) [inline player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js]: `videoElement ` same here

File: player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+            instance.addYOUTUBEListeners(youtubeInstance);
+            instance.setYoutubeStyles(youtubeInstance);
+            instance.videoPlayer = youtubeInstance;
+            /*youtubeInstance.on('volumechange', function(){
+                (youtubeInstance.muted()) ? EkstepRendererAPI.dispatchEvent('renderer:overlay:mute') : EkstepRendererAPI.dispatchEvent('renderer:overlay:unmute');
+            });*/            
+        });
+    },
+    setYoutubeStyles: function(youtube) {
+        var instance = this;
```
Comment (manjudr) [inline player/public/coreplugins/org.ekstep.videorenderer-1.1/renderer/plugin.js]: `instance ` is required?

**Comment by manjudr** [comment]
@gouravmore  Have you tested in mobile and web after minifying with webpack ?

**Comment by gouravmore** [comment]
Created new PR https://github.com/project-sunbird/sunbird-content-player/pull/67. Hence, closing this PR

**Pull request closed without merging**